bindings/src/main/java/io/pravega/storage/hdfs/HDFSStorage.java

```diff
@@ -378,56 +408,76 @@ public void write(SegmentHandle handle, long offset, InputStream data, int lengt
             IOUtils.copyBytes(data, stream, (long) length, false);
 
             stream.flush();
+
+            HDFSMetrics.WRITE_LATENCY.reportSuccessEvent(timer.getElapsed());
+            HDFSMetrics.WRITE_BYTES.add(length);
+            LoggerHelpers.traceLeave(log, "write", traceId, handle, offset, length);
+        } catch (PathNotFoundException | FileNotFoundException e) {
+            // We reach here only if the file with current epoch is not present or sealed.
+            FileStatus status = null;
+            try {
+                status = findStatusForSegment(handle.getSegmentName(), true);
+                if (isSealed(status.getPath())) {
+                    throw new StreamSegmentSealedException(handle.getSegmentName());
+                }
+                if (getEpochFromPath(status.getPath()) != this.epoch) {
```

Comment:
You only throw this if the file epoch is greater than our epoch

If lower, then you throw `DataCorruptionException` because an external actor renamed the file to have a smaller epoch than the one we established during `openWrite`. You said this yourself in the big comment block above.